From 368a0ea0ac67d76f024def4ce238d4a9011e3980 Mon Sep 17 00:00:00 2001
From: Sachin Prabhu <sprabhu@redhat.com>
Date: Mon, 28 Jul 2014 12:16:25 -0400
Subject: [fs] cifs: Warn if SMB3 encryption required by server

Message-id: <1406549822-10376-33-git-send-email-sprabhu@redhat.com>
Patchwork-id: 86742
O-Subject: [RHEL7.1 PATCH 32/69] [bz1082049][CIFS] Warn if SMB3 encryption required by server
Bugzilla: 1082049
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Steve French <smfrench@gmail.com>

We do not support SMB3 encryption yet, warn if server responds
that SMB3 encryption is mandatory.

Signed-off-by: Steve French <smfrench@gmail.com>
(cherry picked from commit 0cbaa53cdd33080c1e2d67ad9295b83c7954f2b3)

Signed-off-by: Sachin Prabhu <sprabhu@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/cifs/smb2pdu.c b/fs/cifs/smb2pdu.c
index 9b0452d..eec06d4 100644
--- a/fs/cifs/smb2pdu.c
+++ b/fs/cifs/smb2pdu.c
@@ -630,6 +630,8 @@ ssetup_ntlmssp_authenticate:
   goto ssetup_exit;
 
  ses->session_flags = le16_to_cpu(rsp->SessionFlags);
+ if (ses->session_flags & SMB2_SESSION_FLAG_ENCRYPT_DATA)
+  cifs_dbg(VFS, "SMB3 encryption not supported yet\n");
 ssetup_exit:
  free_rsp_buf(resp_buftype, rsp);
 
diff --git a/fs/cifs/smb2pdu.h b/fs/cifs/smb2pdu.h
index b50a129..f88320b 100644
--- a/fs/cifs/smb2pdu.h
+++ b/fs/cifs/smb2pdu.h
@@ -122,6 +122,23 @@ struct smb2_pdu {
  __le16 StructureSize2; /* size of wct area (varies, request specific) */
 } __packed;
 
+struct smb2_transform_hdr {
+ __be32 smb2_buf_length; /* big endian on wire */
+    /* length is only two or three bytes - with
+     one or two byte type preceding it that MBZ */
+ __u8   ProtocolId[4]; /* 0xFD 'S' 'M' 'B' */
+ __u8   Signature[16];
+ __u8   Nonce[11];
+ __u8   Reserved[5];
+ __le32 OriginalMessageSize;
+ __u16  Reserved1;
+ __le16 EncryptionAlgorithm;
+ __u64  SessionId;
+} __packed;
+
+/* Encryption Algorithms */
+#define SMB2_ENCRYPTION_AES128_CCM __constant_cpu_to_le16(0x0001)
+
 /*
  * SMB2 flag definitions
  */
@@ -237,6 +254,7 @@ struct smb2_sess_setup_req {
 /* Currently defined SessionFlags */
 #define SMB2_SESSION_FLAG_IS_GUEST 0x0001
 #define SMB2_SESSION_FLAG_IS_NULL 0x0002
+#define SMB2_SESSION_FLAG_ENCRYPT_DATA 0x0004
 struct smb2_sess_setup_rsp {
  struct smb2_hdr hdr;
  __le16 StructureSize; /* Must be 9 */
-- 
1.7.1